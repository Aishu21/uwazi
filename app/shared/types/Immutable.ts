/** @format */

export type IImmutable<T> = T extends string
  ? string
  : T extends number
  ? number
  : T extends boolean
  ? boolean
  : T extends null
  ? null
  : T extends undefined
  ? undefined
  : T extends Array<infer Elem>
  ? {
      toJS(): Elem[];
      get(i: number): IImmutable<Elem>;
    }
  : {
      toJS(): T;
      get<Field extends keyof T>(_field: Field): IImmutable<T[Field]>;
    };
