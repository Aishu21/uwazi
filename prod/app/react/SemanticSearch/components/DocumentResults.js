"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.default = exports.DocumentResults = void 0;
var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _redux = require("redux");
var _reactRedux = require("react-redux");
var _immutable = _interopRequireDefault(require("immutable"));

var _I18N = require("../../I18N");
var _SnippetList = _interopRequireDefault(require("../../Documents/components/SnippetList"));
var _uiActions = require("../../Viewer/actions/uiActions");
var _reactReduxForm = require("react-redux-form");
var _ReactReduxForms = require("../../ReactReduxForms");
var _Icon = require("../../Layout/Icon");
var _Layout = require("../../Layout");function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}function _interopRequireWildcard(obj) {if (obj && obj.__esModule) {return obj;} else {var newObj = {};if (obj != null) {for (var key in obj) {if (Object.prototype.hasOwnProperty.call(obj, key)) {var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {};if (desc.get || desc.set) {Object.defineProperty(newObj, key, desc);} else {newObj[key] = obj[key];}}}}newObj.default = obj;return newObj;}}var REACT_ELEMENT_TYPE;function _jsx(type, props, key, children) {if (!REACT_ELEMENT_TYPE) {REACT_ELEMENT_TYPE = typeof Symbol === "function" && Symbol["for"] && Symbol["for"]("react.element") || 0xeac7;}var defaultProps = type && type.defaultProps;var childrenLength = arguments.length - 3;if (!props && childrenLength !== 0) {props = { children: void 0 };}if (props && defaultProps) {for (var propName in defaultProps) {if (props[propName] === void 0) {props[propName] = defaultProps[propName];}}} else if (!props) {props = defaultProps || {};}if (childrenLength === 1) {props.children = children;} else if (childrenLength > 1) {var childArray = new Array(childrenLength);for (var i = 0; i < childrenLength; i++) {childArray[i] = arguments[i + 3];}props.children = childArray;}return { $$typeof: REACT_ELEMENT_TYPE, type: type, key: key === undefined ? null : '' + key, ref: null, props: props, _owner: null };}

const findResultsAboveThreshold = (results, threshold) => {
  const boundingIndex = results.findIndex(({ score }) => score < threshold);
  return boundingIndex >= 0 ? results.slice(0, boundingIndex) : results;
};

const getSnippetsFromResults = (results, template) => {
  const snippets = results.reduce((_memo, item) => {
    const memo = _memo;
    const text = `${item.text} (${(item.score * 100).toFixed(2)}%)`;
    memo.count += 1;
    if (isNaN(Number(item.page))) {
      const field = template.get('properties').find(p => p.get('name') === item.page).get('label');
      memo.metadata[field] = (memo.metadata[field] || []).concat([text]);
      return memo;
    }
    memo.fullText.push({
      page: Number(item.page),
      text });

    return memo;
  }, { count: 0, metadata: {}, fullText: [] });

  snippets.metadata = Object.keys(snippets.metadata).
  map(field => ({ field, texts: snippets.metadata[field] }));
  return snippets;
};

class DocumentResults extends _react.Component {
  renderSnippetsList(doc, snippets, documentViewUrl) {
    return (
      _jsx(_SnippetList.default, {
        doc: _immutable.default.fromJS(doc),
        documentViewUrl: documentViewUrl,
        snippets: snippets,
        searchTerm: "",
        selectSnippet: (...args) => {this.props.selectSnippet(...args);} }));


  }

  renderFilter() {
    const { threshold } = this.props;
    return (
      _jsx("dl", { className: "metadata-type-text" }, void 0,
      _jsx("dt", { className: "item-header" }, void 0,
      _jsx(_I18N.Translate, {}, void 0, "Threshold"), " ", (threshold * 100).toFixed(2), " %"),

      _jsx("dd", {}, void 0,
      _jsx(_reactReduxForm.Form, { model: "semanticSearch.resultsFilters" }, void 0,
      _jsx(_ReactReduxForms.NumericRangeSlide, {
        model: ".threshold",
        min: 0.3,
        max: 1,
        step: 0.01,
        delay: 200,
        minLabel: (0, _I18N.t)('System', 'Exploration'),
        maxLabel: (0, _I18N.t)('System', 'Precision') })))));





  }

  render() {
    const { doc, threshold, template } = this.props;

    if (!doc.semanticSearch) {
      return false;
    }
    const filteredResults = findResultsAboveThreshold(doc.semanticSearch.results, threshold);
    const snippets = _immutable.default.fromJS(getSnippetsFromResults(filteredResults, template));
    const documentViewUrl = doc.file ? `/document/${doc.sharedId}` : `/entity/${doc.sharedId}`;

    return (
      _jsx(_react.default.Fragment, {}, void 0,
      _jsx("div", { className: "view" }, void 0,
      _jsx("div", { className: "item-info" }, void 0,
      _jsx("div", {}, void 0,
      _jsx(_Icon.Icon, { className: "item-icon item-icon-center", data: doc.icon }),
      _jsx("h1", { className: "item-name" }, void 0,
      doc.title,
      _jsx(_Layout.DocumentLanguage, { doc: _immutable.default.fromJS(doc) }))),


      _jsx(_Layout.TemplateLabel, { template: doc.template })),

      this.renderFilter(),
      _jsx("dl", { className: "metadata-type-numeric" }, void 0,
      _jsx("dt", {}, void 0, _jsx(_I18N.Translate, {}, void 0, "Number of sentences above threshold")),
      _jsx("dd", {}, void 0, filteredResults.length)),

      _jsx("dl", { className: "metadata-type-numeric" }, void 0,
      _jsx("dt", {}, void 0, _jsx(_I18N.Translate, {}, void 0, "% of sentences above threshold")),
      _jsx("dd", {}, void 0, (filteredResults.length / doc.semanticSearch.totalResults * 100).toFixed(2), "%"))),


      this.renderSnippetsList(doc, snippets, documentViewUrl)));


  }}exports.DocumentResults = DocumentResults;


DocumentResults.defaultProps = {
  template: undefined };









const mapStateToProps = ({ semanticSearch, templates }) => ({
  threshold: semanticSearch.resultsFilters.threshold,
  template: templates.find(tpl => tpl.get('_id') === semanticSearch.selectedDocument.get('template')) });



function mapDispatchToProps(dispatch) {
  return (0, _redux.bindActionCreators)({ selectSnippet: _uiActions.selectSnippet }, dispatch);
}var _default =

(0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(DocumentResults);exports.default = _default;