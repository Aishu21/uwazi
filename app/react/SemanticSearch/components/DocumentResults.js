import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import Immutable from 'immutable';

import { Translate, t } from 'app/I18N';
import SnippetList from 'app/Documents/components/SnippetList';
import { selectSnippet } from 'app/Viewer/actions/uiActions';
import { Form } from 'react-redux-form';
import { NumericRangeSlide } from 'app/ReactReduxForms';
import { Icon } from 'app/Layout/Icon';
import { TemplateLabel, DocumentLanguage } from 'app/Layout';

const findResultsAboveThreshold = (results, threshold) => {
  const boundingIndex = results.findIndex(({ score }) => score < threshold);
  return results.slice(0, boundingIndex);
};

export class DocumentResults extends Component {
  renderSnippetsList(doc, snippets, documentViewUrl) {
    return (
      <SnippetList
        doc={Immutable.fromJS(doc)}
        documentViewUrl={documentViewUrl}
        snippets={snippets}
        searchTerm=""
        selectSnippet={(...args) => { this.props.selectSnippet(...args); }}
      />
    );
  }

  renderFilter() {
    const { threshold } = this.props;
    return (
      <dl className="metadata-type-text">
        <dt className="item-header">
          <Translate>Threshold</Translate> {threshold * 100} %
        </dt>
        <dd>
          <Form model="semanticSearch.resultsFilters">
            <NumericRangeSlide
              model=".threshold"
              min={0.3}
              max={1}
              step={0.01}
              delay={200}
              minLabel={t('System', 'More exploration')}
              maxLabel={t('System', 'More precision')}
            />
          </Form>
        </dd>
      </dl>
    );
  }

  render() {
    const { doc, threshold } = this.props;
    if (!doc.semanticSearch) {
      return false;
    }
    const filteredResults = findResultsAboveThreshold(doc.semanticSearch.results, threshold).sort((a, b) => a.score < b.score);
    const snippetsToRender = doc.semanticSearch.results.map(s => Object.assign(
      {}, s, { text: `${s.text} (${(s.score * 100).toFixed(2)}%)` })
    );
    const snippets = Immutable.fromJS({ count: snippetsToRender.length, metadata: [], fullText: snippetsToRender });
    const documentViewUrl = doc.file ? `/document/${doc.sharedId}` : `/entity/${doc.sharedId}`;
    return (
      <React.Fragment>
        <div className="view">
          <div className="item-info">
            <div>
              <Icon className="item-icon item-icon-center" data={doc.icon} />
              <h1 className="item-name">
                {doc.title}
                <DocumentLanguage doc={Immutable.fromJS(doc)} />
              </h1>
            </div>
            <TemplateLabel template={doc.template}/>
          </div>
          {this.renderFilter()}
          <dl className="metadata-type-numeric">
            <dt><Translate>Sentences above threshold</Translate></dt>
            <dd>{ doc.semanticSearch.numRelevant }</dd>
          </dl>
          <dl className="metadata-type-numeric">
            <dt><Translate>% of document above threshold</Translate></dt>
            <dd>{ (doc.semanticSearch.relevantRate * 100).toFixed(2) }%</dd>
          </dl>
        </div>
        {this.renderSnippetsList(doc, snippets, documentViewUrl)}
      </React.Fragment>
    );
  }
}

DocumentResults.propTypes = {
  doc: PropTypes.object.isRequired,
  threshold: PropTypes.number.isRequired,
  selectSnippet: PropTypes.func.isRequired,
};

const mapStateToProps = ({ semanticSearch }) => ({
    threshold: semanticSearch.resultsFilters.threshold
});


function mapDispatchToProps(dispatch) {
  return bindActionCreators({ selectSnippet }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(DocumentResults);
