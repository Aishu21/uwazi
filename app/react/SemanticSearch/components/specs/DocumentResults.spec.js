import React from 'react';
import { DocumentResults } from '../DocumentResults';
import { shallow } from 'enzyme';
import Immutable from 'immutable';

describe('DocumentResults', () => {
  let props;
  let component;
  beforeEach(() => {
    props = {
      doc: Immutable.fromJS({}),
      filters: { threshold: 10 },
      selectTab: jasmine.createSpy('selectTab'),
      selectSnippet: jasmine.createSpy('selectSnippet')
    };

    component = shallow(<DocumentResults {...props}/>);
  });

  describe('render', () => {
    it('should render a result', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
