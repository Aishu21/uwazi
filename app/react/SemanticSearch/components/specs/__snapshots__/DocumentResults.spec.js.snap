// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentResults render should render a result 1`] = `
<React.Fragment>
  <div
    className="view"
  >
    <div
      className="item-info"
    >
      <div>
        <Icon
          className="item-icon item-icon-center"
        />
        <h1
          className="item-name"
        >
          <Connect(DocumentLanguage)
            doc={
              Immutable.Map {
                "semanticSearch": Immutable.Map {
                  "results": Immutable.List [
                    Immutable.Map {
                      "score": 9,
                    },
                    Immutable.Map {
                      "score": 11,
                    },
                  ],
                },
                "avgScore": 10,
              }
            }
          />
        </h1>
      </div>
      <Connect(TemplateLabel) />
    </div>
    <dl
      className="metadata-type-text"
    >
      <dt
        className="item-header"
      >
        <Connect(Translate)>
          Threshold
        </Connect(Translate)>
         
        30
         %
      </dt>
      <dd>
        <Connect(Form)
          model="semanticSearch.resultsFilters"
        >
          <NumericRangeSlide
            delay={200}
            max={1}
            min={0.3}
            model=".threshold"
            step={0.01}
          />
        </Connect(Form)>
      </dd>
    </dl>
    <dl
      className="metadata-type-numeric"
    >
      <dt>
        <Connect(Translate)>
          Sentences above threshold
        </Connect(Translate)>
      </dt>
      <dd>
        1
      </dd>
    </dl>
    <dl
      className="metadata-type-numeric"
    >
      <dt>
        <Connect(Translate)>
          % of document above threshold
        </Connect(Translate)>
      </dt>
      <dd>
        50.00
        %
      </dd>
    </dl>
  </div>
  <Connect(Component)
    doc={
      Immutable.Map {
        "semanticSearch": Immutable.Map {
          "results": Immutable.List [
            Immutable.Map {
              "score": 9,
            },
            Immutable.Map {
              "score": 11,
            },
          ],
        },
        "avgScore": 10,
      }
    }
    documentViewUrl="/entity/undefined"
    searchTerm=""
    selectSnippet={[Function]}
    snippets={
      Immutable.Map {
        "count": 1,
        "metadata": Immutable.List [],
        "fullText": Immutable.List [
          Immutable.Map {
            "score": 9,
            "text": "undefined (900.00%)",
          },
        ],
      }
    }
  />
</React.Fragment>
`;
